adding tfa2017 protocol with unit test

behrisch opened the pull request:
The TFA protocol with a unit test against the rewrite branch.

CurlyMoo commented:
Can you sync with the latest code so the unittest actually succeeds?

behrisch commented:
The test which fails is not the one you fixed, so I think it won't help.

CurlyMoo commented on libs/pilight/protocols/433.92/tfa2017.c:
```diff
@@ -0,0 +1,246 @@
+	options_add(&tfa2017->options, "0", "show-humidity", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "^[10]{1}$");
+	options_add(&tfa2017->options, "0", "show-temperature", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "^[10]{1}$");
+
+
```
Can you remove this bogus new line?

behrisch replied:
OK, can I add commits to this PR or do you want a new one which is squashed?

CurlyMoo replied:
Just push to the current branch. Sqaushing can be done in git when merging.

CurlyMoo commented on libs/pilight/protocols/433.92/tfa2017.c:
```diff
@@ -0,0 +1,246 @@
+void tfa2017Init(void) {
+	protocol_register(&tfa2017);
+	protocol_set_id(tfa2017, "tfa2017");
+	protocol_device_add(tfa2017, "tfa2017", "TFA 30.X Temp Hum Sensor Protocol 2017");
```
Is the addition of protocol `Protocol 2017` necessary?

behrisch replied:
I just wanted to give it a description different from the string which is in tfa30. Suggestions are welcome.

CurlyMoo replied:
Maybe `tfa3.0-2`?
And `TFA 30.X Temp Hum Sensor Protocol v2`?

behrisch replied:
I am not sure, I think 3.0 is misleading because as far as I understood the 30 refers to the first two digits of the product number (Kat.-Nr.?) as printed on the device which is 30.3208.02 for my sender and 30.3054.10 on the receiver. I will ask in the forum threads maybe there is a good idea for a name and description.

CurlyMoo commented on tests/protocols/tfa2017.h:
```diff
@@ -0,0 +1,15 @@
+static const int tfa2017_nrtests = 2;
+static const struct raw_t tfa2017_tests[] = {
+	{
+		"4752 1032 912 1039 436 613 841 594 373 1056 431 532 443 535 436 537 927 552 415 558 418 561 416 1037 920 1033 922 555 420 557 416 1034 923 557 424 546 436 1014 453 516 944 535 435 546 444 998 473 506 944 1012 943 1007 475 496 962 526 446 526 446 996 486 490 488 489 962 517 465 516 457 514 466 510 465 985 495 497 490 472 498 476 504 488 491 473 495 475 507 467 509 473 505 465 507 496 961 973 97 ...
```
Maybe add a test with a negative temperature?

behrisch replied:
Will do as soon as it is cold enough, because all my tests are based on real measurements.

behrisch commented:
I added more tests and removed the one line. No answers on the forum about the name though (and there is again one unit tests failing which seems to be non-deterministic).

CurlyMoo commented on libs/pilight/protocols/433.92/tfa2017.c:
```diff
@@ -130,9 +130,12 @@ static void parseCode(char **message) {
 		}
 		tmp = tmp->next;
 	}
+	if (humidity<0. || humidity>100.) {
```
Can you make this:
```
if(humidity < 0 || humidity > 100) {
```

CurlyMoo commented on libs/pilight/protocols/433.92/tfa2017.c:
```diff
@@ -0,0 +1,249 @@
+	}
+	// The protocol sends the message three times in a row.
+	// If we find two identical ones, we consider it valid.
+	if (s<2) {
```
Please check the pilight style guide.

CurlyMoo commented on tests/protocols/tfa2017.h:
```diff
@@ -1,14 +1,32 @@
+		NULL,
+		0
+	},
+	{ // error raw length too short
```
Nice

CurlyMoo commented:
All checks succeeded. Good work.

behrisch commented:
Thanks. I hope it will be useful.

CurlyMoo commented:
Can you change the license to MPLv2?

CurlyMoo commented on libs/pilight/protocols/433.92/tfa2017.c:
```diff
@@ -1,19 +1,9 @@
 /*
 	Copyright (C) 2014, 2018 CurlyMo & DonBernos & Michael Behrisch
```
And of course make sure the indenting matches :smile:

CurlyMoo commented:
Can you fix this warning?
https://travis-ci.org/pilight/pilight/builds/478696545#L901
```
 snprintf(*message, 255, "");
```

change it to:
```
memset(*message, 0, 255);
```
or maybe:
```
memset(&(*message), 0, 255);
```

CurlyMoo commented:
Ready for merging?

behrisch commented:
I am, if you are :-)

CurlyMoo commented:
Thanks for this wonderful merge request!

behrisch commented:
Thank you for integrating! I had a short look at the checksum code in https://github.com/adilosa/weathermon/blob/master/weathermon.c which may be adaptable to this protocol but my first try did not work out.
